### PR TITLE
Remove External Connection Secrets from operator responsibilities

### PR DESCRIPTION
**Pull request by gmfrasca**
Resolves https://github.com/opendatahub-io/data-science-pipelines-operator/issues/151

DSPO should not be responsible for maintaining the connections secrets for an external database or object store.  Remove this from the maintained artifacts list.

## Description
DSPO should not be responsible for maintaining the connections secrets for an external database or object store.  Remove this from the maintained artifacts list.

## How Has This Been Tested?
For all tests, assume DSPA is named `sample`

There are essentially 9 test cases/DSPAs here:

** Default Values Test **
- both Default MariaDB and Minio options [expected: will create `ds-pipline-db-sample` and `ds-pipeline-s3-sample` secrets]

** Object Storage Tests **
- DSPA with `spec.objectStorage.minio.s3CredentialsSecret` specified but the secret specified in `secretName` does not exist yet [expected: DSPO will create this secret and generate random credentials for the accesskey and secretkey]
- DSPA with `spec.objectStorage.minio.s3CredentialsSecret` specified and the secret specified in `secretName` already exists [expected: DSPO will create Minio pod but reuse values found in this secret]
- DSPA with `spec.objectStorage.externalStorage.s3CredentialsSecret` specified but the secret specified in `secretName` does not exist yet [expected: DSPO will not spin up any pods, and will return an error]
- DSPA with `spec.objectStorage.externalStorage.s3CredentialsSecret` specified and the secret specified in `secretName` already exists [expected: DSPO will spin up a DSPA and successfully connect to the External S3 Storage]

** Database Tests **
- DSPA with `spec.database.mariaDB.passwordSecret` specified but the secret specified in `name` does not exist yet [expected: DSPO will create this secret and generate random credentials for the password]
- DSPA with `spec.database.mariaDB.passwordSecret` specified and the secret specified in `name` already exists [expected: DSPO will create MariaDB pod but reuse the password found in this secret]
- DSPA with `spec.database.externalDB.passwordSecret` specified but the secret specified in `name` does not exist yet [expected: DSPO will not spin up any pods, and will return an error]
- DSPA with `spec.database.externalDB.passwordSecret` specified and the secret specified in `name` already exists [expected: DSPO will spin up a DSPA and successfully connect to the External Database]

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by gmfrasca** [comment]
**Secrets used for testing:**

ds-pipeline-db-sample:
```yaml
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: ds-pipeline-db-sample
data:
  password: {{ base64-encoded Database password }}
```

mlpipeline-minio-artifact:
```yaml
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: mlpipeline-minio-artifact
data:
  accesskey: {{ base64-encoded S3 Access Key }}
  host: {{ base64-encoded S3 Endpoint URL }}
  port: {{ base64-encoded S3 Endpoint Port }}
  secretkey: {{ base64-encoded S3 Secret Key }}
  secure: {{ base64-encoded boolean true/false (true if https, false if http) }}
```

File: controllers/storage.go [hunk 1]
```diff
@@ -22,13 +22,10 @@ import (
 	dspav1alpha1 "github.com/opendatahub-io/data-science-pipelines-operator/api/v1alpha1"
 )
 
-const storageSecret = "minio/secret.yaml.tmpl"
-
 var storageTemplates = []string{
 	"minio/deployment.yaml.tmpl",
 	"minio/pvc.yaml.tmpl",
 	"minio/service.yaml.tmpl",
-	storageSecret,
```
Comment (HumairAK) [inline controllers/storage.go:31]: This will break deployments that are using minio (so no minio specified), as it will not deploy the randomly generated secret. Even though minio is unsupported, we should consider keeping this around to for simplifying minio deployments.

**Comment by HumairAK** [comment]
Similarly for [this](https://github.com/gmfrasca/data-science-pipelines-operator/blob/rm-secret-creation/config/internal/apiserver/deployment.yaml.tmpl#L120-L131), you will likely need to update [this](https://github.com/gmfrasca/data-science-pipelines-operator/blob/rm-secret-creation/controllers/dspipeline_params.go#L220)

File: controllers/database.go [hunk 2]
```diff
@@ -21,14 +21,11 @@ import (
...
 )
 
-const dbSecret = "mariadb/secret.yaml.tmpl"
-
 var dbTemplates = []string{
 	"mariadb/deployment.yaml.tmpl",
 	"mariadb/pvc.yaml.tmpl",
 	"mariadb/sa.yaml.tmpl",
 	"mariadb/service.yaml.tmpl",
-	dbSecret,
```
[the remark this answers is not in the document]
Comment (gmfrasca) [reply controllers/database.go:31]: wasn't that the intent of this ask?  That DSPO should never make secrets even if it's default (this is why I included/migrated the functionality to this idea of DevTools)

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`

File: controllers/dspipeline_params.go [hunk 3]
```diff
@@ -122,28 +122,71 @@ func passwordGen(n int) string {
...
 		// Assume validation for CR ensures these values exist
 		p.DBConnection.Host = dsp.Spec.Database.ExternalDB.Host
 		p.DBConnection.Port = dsp.Spec.Database.ExternalDB.Port
 		p.DBConnection.Username = dsp.Spec.Database.ExternalDB.Username
 		p.DBConnection.DBName = dsp.Spec.Database.ExternalDB.DBName
-		customCreds = dsp.Spec.Database.ExternalDB.PasswordSecret
+		p.DBConnection.CredentialsSecret = dsp.Spec.Database.ExternalDB.PasswordSecret
+
+		// Retreive DB Password from specified secret.  Handle errors (empty password) later
+		password, _ := p.RetrieveSecret(ctx, client, p.DBConnection.CredentialsSecret.Name, p.DBConnection.CredentialsSecret.Key, log)
```
Comment (HumairAK) [inline controllers/dspipeline_params.go:188]: not sure how I feel about dropping the err like this, we should handle it properly if an error is encountered, if the resource is not found handle it when handling the error via `apierrs.IsNotFound(err)`

currently we could be hiding other forms of errors we may not necessarily be aware of, so we should expose them

File: controllers/dspipeline_params.go [hunk 4]
```diff
@@ -174,59 +218,25 @@ func (p *DSPAParams) SetupDBParams(ctx context.Context, dsp *dspa.DataSciencePip
...
-	createNewSecret := false
 
-	// Attempt to fetch the specified DB secret
-	err := client.Get(ctx, namespacedName, dbSecret)
-	if err != nil && apierrs.IsNotFound(err) {
-		if !customCredentialsSpecified {
-			generatedPass := passwordGen(12)
-			p.DBConnection.Password = base64.StdEncoding.EncodeToString([]byte(generatedPass))
-			createNewSecret = true
+		// If custom DB Secret provided, use its values.  Otherwise generate a default
```
Comment (HumairAK) [inline controllers/dspipeline_params.go:222]:
```suggestion
		// If custom DB Secret provided, use its values. Otherwise, generate a default
```

File: controllers/dspipeline_params.go [hunk 5]
```diff
@@ -265,7 +265,13 @@ func (p *DSPAParams) SetupObjectParams(ctx context.Context, dsp *dspa.DataScienc
 
 		// Port can be empty, which is fine.
 		p.ObjectStorageConnection.Port = dsp.Spec.ObjectStorage.ExternalStorage.Port
-		customCreds = dsp.Spec.ObjectStorage.ExternalStorage.S3CredentialSecret
+		p.ObjectStorageConnection.CredentialsSecret = dsp.Spec.ObjectStorage.ExternalStorage.S3CredentialSecret
+
+		// Retrieve ObjStore Creds from specified secret.  Handle issues (empty creds) later.
+		accesskey, _ := p.RetrieveSecret(ctx, client, p.ObjectStorageConnection.CredentialsSecret.SecretName, p.ObjectStorageConnection.CredentialsSecret.AccessKey, log)
+		secretkey, _ := p.RetrieveSecret(ctx, client, p.ObjectStorageConnection.CredentialsSecret.SecretName, p.ObjectStorageConnection.CredentialsSecret.SecretKey, log)
```
Comment (HumairAK) [inline controllers/dspipeline_params.go:272]: same as earlier, we should expose the error and not make the user guess what it is

**Review by HumairAK — CHANGES_REQUESTED** [review]

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`

**Comment by HumairAK** [comment]
currently this doesn't work for s3 storage because we are not passing in `OBJECTSTORECONFIG_CREDENTIALSSECRET` to the api server, so I think you'll need to do that. right now it's still expecting the default secret so pipelines are failing with `Error: secret "mlpipeline-minio-artifact" not found`

https://github.com/opendatahub-io/data-science-pipelines/commit/8d492002883653d4b93c2f2978519a24be1473f6#diff-47d6ab6bca869e0559465170316707ceda42f65103bd0d4e93519294305e62d1R50

File: controllers/dspipeline_params.go [hunk 6]
```diff
@@ -174,59 +222,25 @@ func (p *DSPAParams) SetupDBParams(ctx context.Context, dsp *dspa.DataSciencePip
...
+			p.DBConnection.CredentialsSecret = p.MariaDB.PasswordSecret
 		} else {
-			log.Error(err, fmt.Sprintf("DB secret [%s] was specified in CR but does not exist.",
-				credsSecretName))
+			p.DBConnection.CredentialsSecret = &dspa.SecretKeyValue{
+				Name: config.DefaultDBSecretNamePrefix + p.Name,
+				Key:  config.DefaultDBSecretKey,
+			}
+		}
+		dbPassword, err := p.RetrieveOrCreateDBSecret(ctx, client, p.DBConnection.CredentialsSecret, log)
```
Comment (HumairAK) [inline controllers/dspipeline_params.go:235]: Is the expected behavior for: 

```yaml
spec:
  database:
    mariaDB:
      passwordSecret:
        name: a-custom-db-secret
        key: randomkey
```

that `a-custom-db-secret` is always created? if I provide `a-custom-db-secret` before deploying the dspa, should it not utilize that secret? currently it seems like it will always overwrite this secret

```bash

$ cat dsp.yaml 

apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: custom-default-secret
spec:
  database:
    mariaDB:
      passwordSecret:
        name: a-custom-db-secret
        key: randomkey
  objectStorage:
    minio:
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: 'quay.io/opendatahub/odh-ml-pipelines-frontend-container:beta-ui'
---
apiVersion: v1
kind: Secret
metadata:
  name: a-custom-db-secret
stringData:
  randomkey: customDefaultDBSecretTest

$ oc apply -f dspa.yaml 

$ oc get secret a-custom-db-secret -o yaml | yq .data.randomkey | base64 -d  && echo
SJ3Lm6RtY7MB


$


```

Comment (HumairAK) [reply controllers/dspipeline_params.go:235]: same for object storage

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`

File: controllers/dspipeline_params.go (hunk 6 above)
Comment (gmfrasca) [reply controllers/dspipeline_params.go:235]: It should *not* overwrite a pre-existing secret (hence the Retrieve in RetrieveOrCreateSecret).  Do you have logs showing when this happens, the code shouldn't allow it

Comment (gmfrasca) [reply controllers/dspipeline_params.go:235]: This is the code that handles it.  It should only generate a password if we attempt to retrieve the secret and it returns a NotFound error.  any other case either uses the retrieved value or rethrows an error
```go
val, err := p.RetrieveSecret(ctx, client, secretName, secretKey, log)
	if err != nil && apierrs.IsNotFound(err) {
		generatedPass := passwordGen(generatedPasswordLength)
		return base64.StdEncoding.EncodeToString([]byte(generatedPass)), nil
	} else if err != nil {
		log.Error(err, "Unable to create DB secret...")
		return "", err
	}
	log.Info(fmt.Sprintf("Secret [%s] already exists, using stored value.", secretName))
	return val, nil
}
```

Comment (gmfrasca) [reply controllers/dspipeline_params.go:235]: ah. you are apply these simultaneously.  I wonder if this is a race condition (DSPO detects the secret doesn't exist before the secret was created.  maybe try apply them seperately with a small wait (secret first, then DSPA)

Comment (HumairAK) [reply controllers/dspipeline_params.go:235]: ooo good point, I'll try that

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`

**Comment by HumairAK** [comment]
I have pretty thoroughly tested this and logged the outcomes: 
Check marks indicate the expected outcome, unchecks indicate it not meet the outcome. 

The default cases I don't really care that much, they seem like more edge cases.

---

Verification check list:

Standard Test suite (STS): 
- Simple DSPA deployed successfully
- Run Iris Pipeline Successfully
- Check S3 storage for artifacts passing/storage
- Retrieve completed run (db test)


For the `DSPA + Secret at the same time` deploy cases, the Secrets are listed after the DSPA in the same yaml file.

<details>

<summary>sample.yaml</summary>

```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: case3
  namespace: dspa
spec:
  apiServer:
    image: quay.io/opendatahub/ds-pipelines-api-server@sha256:650f196f212664c5d4bd9a293fcd5fd0dca40e1d5c3af67f57b2e69a436070b2
  database:
    mariaDB:
      passwordSecret:
        key: case3dbpsw
        name: case3-dbsecret
  objectStorage:
    minio:
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: 'quay.io/opendatahub/odh-ml-pipelines-frontend-container:beta-ui'
---
kind: Secret
apiVersion: v1
metadata:
  name: case3-dbsecret
stringData:
  case3dbpsw: 
type: Opaque
```


</details>


Default Connections
1. Deploy without any secrets specified
   - [x] Case: Standard Test suite
   - [x] Case: All secrets cleaned up on DSPA delete
2. Deploy custom DB secret specified, Secret first, then DSPA
   - [x] Case: Standard Test suite
   - [ ] Case: DB secret NOT cleaned up on DSPA delete
   - [ ] Case: DB Secret is specified in DSPA, but not deployed, should throw error
3. Deploy custom DB secret specified, DSPA + Secret at the same time
   - [ ] Case: Standard Test suite
4. Deploy custom S3 secret specified, Secret first, then DSPA
   - Constraints: s3 secret name != `mlpipeline-minio-artifact`
   - [x] Case: Standard Test suite
   - [ ] Case: S3 secret NOT cleaned up on DSPA delete
   - [ ] Case: S3 Secret is specified in DSPA, but not deployed, should throw error
5. Deploy custom S3 secret specified, DSPA + Secret at the same time
   - Constraints: s3 secret name != `mlpipeline-minio-artifact`
   - [ ] Case: Standard Test suite
6. Deploy custom S3 AND default DB secret specified, Secret first, then DSPA
   - Constraints: s3 secret name != `mlpipeline-minio-artifact`
   - [x] Case: Standard Test suite

External Connections
1. Deploy using S3/DB external with custom secrets, Secret first, then DSPA
   - Constraints: s3 secret name != `mlpipeline-minio-artifact`
   - [x] Case: Standard Test suite
   - [x] Case: Secrets NOT cleaned up on DSPA delete
2. Deploy using S3/DB external with custom secrets, DSPA + Secret at the same time
   - Constraints: s3 secret name != `mlpipeline-minio-artifact`
   - [x] Case: Standard Test suite
   - [x] Case: Secrets NOT cleaned up on DSPA delete

   > Note this case results in the DSPO operator blocking on a thread for a couple of minutes
   > It then deploys fine.
   ```log
   2023-10-04T19:26:19Z	DEBUG	Unable to retrieve secret [dbsecret].
   2023-10-04T19:26:19Z	INFO	Encountered error when parsing CR: [DB Password from secret [dbsecret] for key [customkey] was not successfully retrieved, ensure that the secret with this key exist.]	{"namespace": "dspa", "dspa_name": "case2"}
   2023-10-04T19:26:19Z	DEBUG	DataSciencePipelinesApplication Reconciler called.	{"namespace": "dspa", "dspa_name": "case2"}
   2023-10-04T19:26:19Z	DEBUG	Unable to retrieve secret [dbsecret].
   2023-10-04T19:26:19Z	INFO	Encountered error when parsing CR: [DB Password from secret [dbsecret] for key [customkey] was not successfully retrieved, ensure that the secret with this key exist.]	{"namespace": "dspa", "dspa_name": "case2"}
   2023-10-04T19:28:19Z	DEBUG	DataSciencePipelinesApplication Reconciler called.	{"namespace": "dspa", "dspa_name": "case2"}
   ```

3. Deploy using S3/DB external with no secrets
   - [x] Case: Simple DSPA fails to deploy
   > Note this does block a thread ~2min, but with a maxconcurrency count > 1, this is fine
   > unless n dspas are blocked, and maxconcurrencycount == n

Upgrade testing: 

Steps (with default MariaDB): 
- Install ODH (v1.8)
- Install DSPA Kfdef with ODH-Dashboard & DS Pipelines (main)
- Create Data Connection + Pipeline Server
- Upgrade Kfdef to leverage custom secrets change in DSPO
- [x] Confirm Pipelines run successfully
- [x] Reboot API Server to confirm connection still works (repull secrets into pod)
- [x] Confirm Pipelines run successfully
- [x] Confirm artifact storage works in iris example
- [x] Confirm Runs are stored in db
- [x] Deleted old minio artifact secret
- [x] Restarted API Server pod
- [x] Confirm Runs still work, artifacts are stored in s3
- [x] Confirm deleting pipeline server does not delete data connection secret

Steps (with External DB):
- Install ODH (v1.8)
- Install DSPA Kfdef with ODH-Dashboard & DS Pipelines (main)
- Create Data Connection + Pipeline Server
- Upgrade Kfdef to leverage custom secrets change in DSPO
- [x] Confirm Pipelines run successfully
- [x] Reboot API Server to confirm connection still works (repull secrets into pod)
- [x] Confirm Pipelines run successfully
- [x] Confirm artifact storage works in iris example
- [x] Confirm Runs are stored in db
- [x] Deleted old minio artifact secret
- [x] Restarted API Server pod
- [x] Confirm Runs still work, artifacts are stored in s3
- [x] Confirm deleting pipeline server does not delete data connection secret
- [ ] Confirm deleting pipeline server does not delete db secret (created by dashboard)

**Comment by HumairAK** [comment]
We want this check: 
```
 Confirm deleting pipeline server does not delete db secret (created by dashboard)
```

@gmfrasca I did not confirm the above for the MariaDB case. I'm not sure if ODH Dashboard creates any secrets for the MariaDB default deployment, please confirm this and if dashboard indeed uses their own secret, we need to handle this too: 

```
Case: DB secret NOT cleaned up on DSPA delete
Case: DB Secret is specified in DSPA, but not deployed, should throw error (not crash)
```

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-118`
